### ISSUE TITLE
State changes aren't being coherently detected/reflected in the UI or the debug output.

### ISSUE THREAD

**Issue by dicoda** [pointfreeco/swift-composable-architecture#337]
**Describe the bug**
State changes are not being detected/reflected in the UI and are not being recognised coherently in the debug console output.

**To Reproduce**

It's the same repo as the swift-composable-architecture, but it has one additional project that started as an exact duplicate of the Todos Demo. 

Steps to reproduce:  
1. build and run
2. add 1 Register to the list, by touching the `Add Register` button
3. Press the checkbox of the newly added item in the list twice. (No changes in the UI for some reason)
4. Look for the console output, it should be printing the changes in the model correctly, but at the same time it says (No state changes).

**Expected behavior**
To detect a change in the state and for that state to be reflected in the UI. 
When I press a checkbox, it should toggle the UI accordingly.

**Screenshots**
I'm using the console output, hope that's ok:

Here's what debug prints the first time I press a checkbox. Notice that the `isDefault` property is set to `1`, but it also says that there are no changes.

```
received action:
  RegistersAction.register(
    id: 0,
    action: RegisterAction.checkBoxToggled
  )
  (No state changes)

received action:
  RegistersAction.got(
    Result<Action, Failure>.success(
      Action.didFetch(
        objects: [
          <RegisterModel: 0x6000036ec870> (entity: RegisterModel; id: 0x8cae174b2dbeaa18 <x-coredata://D81A2E12-D906-4677-9747-1AFC344BF88C/RegisterModel/p1>; data: {
              id = "B9266739-58B4-42EE-B904-A6BF8256F182";
              isDefault = 1;
              isPreset = 0;
              name = Some;
              note = 60;
              order = 0;
              semitones = 12;
          }),
        ]
      )
    )
  )
  (No state changes)

```

Now, when I press the checkbox a second time, see the `isDefault` property, it changes back to `0`. This is to be expected. However, the console output again states that there are no state changes.

```
received action:
  RegistersAction.register(
    id: 0,
    action: RegisterAction.checkBoxToggled
  )
  (No state changes)

received action:
  RegistersAction.got(
    Result<Action, Failure>.success(
      Action.didFetch(
        objects: [
          <RegisterModel: 0x6000036ec870> (entity: RegisterModel; id: 0x8cae174b2dbeaa18 <x-coredata://D81A2E12-D906-4677-9747-1AFC344BF88C/RegisterModel/p1>; data: {
              id = "B9266739-58B4-42EE-B904-A6BF8256F182";
              isDefault = 0;
              isPreset = 0;
              name = Some;
              note = 60;
              order = 0;
              semitones = 12;
          }),
        ]
      )
    )
  )
  (No state changes)
```

The objects received from the action are being properly assigned to the `registers` property of the reducer's managed state. By adding a break point and drilling down with the debugger it's possible to see the newly updated state. However, the issue still present's itself.

**Environment**
 - Xcode 12.3
 - Swift 5.3.2
 - OS: iOs 14.3

**Additional context**

The purpose of this project is to attempt to replicate the Todos Demo with CoreData, so you'll notice the code is very similar. My gut feeling tells me that the TCA doesn't like reference types, and some funky stuff might be going on in the background, but I'm not sure.

Thank you 🙏 

edit: Just updated the uploaded zip file to fix a compile error

**Comment by dicoda** [comment]
Ok, so I couldn't for the life of me understand how to overcome this, so I just left the NSManagedObject inside the State for each item.

```swift
struct RegisterState: Equatable, Identifiable {
   var model: RegisterModel // < NSManagedObject
  // ...
}
```

So when I call forEach on my reducer, it's iterating over an array of value types and not reference types. I suspect I get a new memory pointer when moving items around but changing the property of an object, when it's a reference type, it doesn't know it changed 🤔

**Comment by dicoda** [comment]
Unrelated: 
Turns out leaving the `RegisterModel` or NSManagedObject as a reference in my `RegisterState` was also a bad idea. As soon as I attempted to test mutations I would up needing to assert against the objects created from CoreData, leaving me without full control over what state is final after a mutation from an action. I'm adopting just structs with an UUID for reference, and map whatever I need from CoreData entities to value type structs.

Related:
Since RegisterModel is Equatable, I still think it might be a bug.. But I'm not sure if it's my lack of understanding 🤔
